src/condition/shi_detector.cpp

```diff
@@ -34,7 +34,105 @@ struct shi_result {
     std::vector<std::string> key_path;
 };
 
-std::optional<shi_result> shi_string_impl(std::string_view resource,
+struct shell_argument_array {
+    static constexpr std::size_t npos = std::string_view::npos;
+
+    explicit shell_argument_array(const ddwaf_object &root)
+    {
+        // Since the type check is performed elsewhere, we don't need to check again
+        auto argc = static_cast<std::size_t>(root.nbEntries);
+        if (argc == 0) {
+            return;
+        }
+
+        // Calculate the final resource length
+        std::size_t resource_len = 0;
+        for (std::size_t i = 0; i < argc; ++i) {
+            const auto &child = root.array[i];
+            if (child.type == DDWAF_OBJ_STRING && child.stringValue != nullptr &&
+                child.nbEntries > 0) {
+                // if the string is valid or non-empty, increase the resource
+                // length + 1 for the extra space when relevant
+                resource_len +=
+                    static_cast<std::size_t>(child.nbEntries) + static_cast<std::size_t>(i > 0);
+            }
+        }
+
+        indices.reserve(argc);
+        resource.reserve(resource_len);
+
+        std::size_t index = 0;
+        for (std::size_t i = 0; i < argc; ++i) {
+            const auto &child = root.array[i];
+            if (child.type != DDWAF_OBJ_STRING || child.stringValue == nullptr ||
+                child.nbEntries == 0) {
+                continue;
+            }
+
+            const std::string_view str{
+                child.stringValue, static_cast<std::size_t>(child.nbEntries)};
+
+            indices.emplace_back(index, index + str.size() - 1);
+
+            index += str.size() + 1;
+
+            if (!resource.empty()) {
+                resource.append(" "sv);
+            }
+            resource.append(str);
+        }
+    }
+
+    std::size_t find(std::string_view str, std::size_t start = 0)
+    {
+        while ((start = resource.find(str, start)) != npos) {
+            auto end = start + str.size() - 1;
+
+            // Ensure that both start and end are within the same argument
```

Comment:
Why run this algorithm on the concatenated string rather than the array from the beginning ?

Reply:
The idea was that searching within a single string might have better cache properties than iterating and searching within multiple different strings, which would result in multiple indirections (array + string). Although I have not benchmarked both options.